src/rpdk/core/contract/resource_client.py

```diff
@@ -215,28 +215,67 @@ def get_metadata(self):
-            )
-            if document_input != document_output:
-                transformed_property = self.transform(prop, input_model)
-                self.update_transformed_property(
-                    prop, transformed_property, input_model
+        for prop in self.property_transform_keys:  # pylint: disable=R1702
+            try:
+                document_input, _path_input, _parent_input = traverse(
+                    input_model, list(prop)[1:]
                 )
+                document_output, _path_output, _parent_output = traverse(
+                    output_model, list(prop)[1:]
+                )
+                if not self.compare(document_input, document_output):
+                    path = "/" + "/".join(prop)
+                    property_transform_value = self.property_transform[path].replace(
+                        '"', '\\"'
+                    )
+                    if "OR" in property_transform_value:
+                        transform_functions = property_transform_value.split("OR")
+                        for transform_function in transform_functions:
+                            transformed_property = self.transform(
+                                transform_function, input_model
+                            )
+                            value = self.convert_type(
+                                document_input, transformed_property
+                            )
+                            if self.compare(value, document_output):
+                                self.update_transformed_property(
+                                    prop, value, input_model
+                                )
+                    else:
+                        transformed_property = self.transform(
+                            property_transform_value, input_model
+                        )
+                        value = self.convert_type(document_input, transformed_property)
+                        self.update_transformed_property(prop, value, input_model)
+
+            except KeyError:
+                pass
 
-    def transform(self, property_path, input_model):
-
-        path = "/" + "/".join(property_path)
-        property_transform_value = json.dumps(self.property_transform[path])
-        # self.property_transform[path].replace('"', '\\"')
+    @staticmethod
+    def convert_type(document_input, transformed_property):
+        if isinstance(document_input, bool):
+            return bool(transformed_property)
+        if isinstance(document_input, int):
+            return int(transformed_property)
+        if isinstance(document_input, float):
+            return float(transformed_property)
+        return transformed_property
 
+    @staticmethod
+    def compare(document_input, document_output):
+        try:
+            if isinstance(document_input, str):
+                regex = r"^{}$".format(document_input)
+                re.compile(regex)
+                return re.match(regex, document_output)
```

Comment:
Some regex nitpicking:

The `r` isn't really needed here. It indicates that `\` should not be treated as a special character, and it has no effect on the part that is in the format.

```
>>> print('^{}$'.format('l1\nl2'))
^l1
l2$
```
If we're targeting python 3.6 or higher, we could use f-strings instead of `.format`, but in this case there isn't a big readability difference:

the re.compile(regex) line does nothing here, as you work with the return value of that if you want to compile a regex.
On the other hand: There are not a lot of reasons to compile a regex that you're only using once, so we don't need it anyway.

Putting all those things together, we cloud replace these three lines with

```
return re.match(f"^{document_input}$", document_output)
```

Note that we assume that document_input is a raw string or has its `\`s escaped.